File: api/src/main/java/org/apache/gravitino/job/SparkJobTemplate.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.job;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import org.apache.commons.lang3.StringUtils;

/**
 * Represents a job template for executing Spark applications. This class extends the JobTemplate
 * class and provides functionality specific to Spark job templates, including the class name, jars,
 * files, archives, and configurations required for the Spark job.
 *
 * <p>Take Spark word count job as an example:
 *
 * <p>className: "org.apache.spark.examples.JavaWordCount" executable
 * "https://example.com/spark-examples.jar" arguments: ["{{input_path}}", "{{output_path}}"]
 * configs: {"spark.master": "local[*]", "spark.app.name": "WordCount"}
 *
 * <p>configs is a map of configuration parameters that will be used by the Spark application. It
 * can be templated by using placeholders like "{{foo_value}}" and "{{bar_value}}". These
 * placeholders will be replaced with actual values when the job is executed.
 *
 * <p>jars, files, and archives are lists of resources that will be used by the Spark application .
 * These resources must be accessible to the Gravitino server, and can be located in the local file
 * system, on a web server (e.g., HTTP, HTTPS, FTP). Distributed file systems like HDFS or S3 will
 * be supported in the future.
 */
public class SparkJobTemplate extends JobTemplate {

  private final String className;

  private final List<String> jars;

  private final List<String> files;

  private final List<String> archives;

  private final Map<String, String> configs;

  /**
   * Constructs a SparkJobTemplate with the specified builder.
   *
   * @param builder the builder containing the configuration for the Spark job template
   */
  protected SparkJobTemplate(Builder builder) {
    super(builder);
    this.className = builder.className;
    this.jars = builder.jars;
    this.files = builder.files;
    this.archives = builder.archives;
    this.configs = builder.configs;
  }

  /**
   * Returns the class name of the Spark application to be executed.
   *
   * @return the class name
   */
  public String className() {
    return className;
  }

  /**
   * Returns the list of JAR files required for the Spark job.
   *
   * @return the list of JAR files
   */
  public List<String> jars() {
    return jars;
  }

  /**
   * Returns the list of files required for the Spark job.
   *
   * @return the list of files
   */
  public List<String> files() {
    return files;
  }

  /**
   * Returns the list of archives required for the Spark job.
   *
   * @return the list of archives
   */
  public List<String> archives() {
    return archives;
  }

  /**
   * Returns the configuration map for the Spark job.
   *
   * @return the configuration map
   */
  public Map<String, String> configs() {
    return configs;
  }

  @Override
  public JobType jobType() {
    return JobType.SPARK;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof SparkJobTemplate)) return false;
    if (!super.equals(o)) return false;

    SparkJobTemplate that = (SparkJobTemplate) o;
    return Objects.equals(className, that.className)
        && Objects.equals(jars, that.jars)
        && Objects.equals(files, that.files)
        && Objects.equals(archives, that.archives)
        && Objects.equals(configs, that.configs);
  }

  @Override
  public int hashCode() {
    return Objects.hash(super.hashCode(), className, jars, files, archives, configs);
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder("\nSparkJobTemplate{\n");
    sb.append("  className='").append(className).append("',\n");

    if (!jars.isEmpty()) {
      sb.append("  jars=[\n");
      jars.forEach(j -> sb.append("    ").append(j).append("\n"));
      sb.append("  ],\n");
    } else {
      sb.append("  jars=[],\n");
    }

    if (!files.isEmpty()) {
      sb.append("  files=[\n");
      files.forEach(f -> sb.append("    ").append(f).append("\n"));
      sb.append("  ],\n");
    } else {
      sb.append("  files=[],\n");
    }

    if (!archives.isEmpty()) {
      sb.append("  archives=[\n");
      archives.forEach(a -> sb.append("    ").append(a).append("\n"));
      sb.append("  ],\n");
    } else {
      sb.append("  archives=[],\n");
    }

    if (!configs.isEmpty()) {
      sb.append("  configs={\n");
      configs.forEach((k, v) -> sb.append("    ").append(k).append(": ").append(v).append(",\n"));
      sb.append("  }\n");
    } else {
      sb.append("  configs={}\n");
    }

    return sb + super.toString() + "}\n";
  }

  /**
   * Creates a new builder for constructing instances of {@link SparkJobTemplate}.
   *
   * @return a new instance of {@link Builder}
   */
  public static Builder builder() {
    return new Builder();
  }

  /** Builder for creating instances of {@link SparkJobTemplate}. */
  public static class Builder extends JobTemplate.BaseBuilder<Builder, SparkJobTemplate> {

    private String className;

    private List<String> jars;

    private List<String> files;

    private List<String> archives;

    private Map<String, String> configs;

    private Builder() {}

    /**
     * Sets the class name of the Spark application to be executed.
     *
     * @param className the class name
     * @return the builder instance for method chaining
     */
    public Builder withClassName(String className) {
      this.className = className;
      return this;
    }

    /**
     * Sets the list of JAR files required for the Spark job.
     *
     * @param jars the list of JAR files
     * @return the builder instance for method chaining
     */
    public Builder withJars(List<String> jars) {
      this.jars = jars;
      return this;
    }

    /**
     * Sets the list of files required for the Spark job.
     *
     * @param files the list of files
     * @return the builder instance for method chaining
     */
    public Builder withFiles(List<String> files) {
      this.files = files;
      return this;
    }

    /**
     * Sets the list of archives required for the Spark job.
     *
     * @param archives the list of archives
     * @return the builder instance for method chaining
     */
    public Builder withArchives(List<String> archives) {
      this.archives = archives;
      return this;
    }

    /**
     * Sets the configuration map for the Spark job.
     *
     * @param configs the configuration map
     * @return the builder instance for method chaining
     */
    public Builder withConfigs(Map<String, String> configs) {
      this.configs = configs;
      return this;
    }

    /**
     * Builds the SparkJobTemplate instance with the specified parameters.
     *
     * @return a new instance of SparkJobTemplate
     */
    @Override
    public SparkJobTemplate build() {
      validate();
      return new SparkJobTemplate(this);
    }

    @Override
    protected Builder self() {
      return this;
    }

    @Override
    protected void validate() {
      super.validate();

      Preconditions.checkArgument(
          StringUtils.isNotBlank(className), "Class name must not be null or empty");

      this.jars = jars != null ? ImmutableList.copyOf(jars) : ImmutableList.of();
      this.files = files != null ? ImmutableList.copyOf(files) : ImmutableList.of();
      this.archives = archives != null ? ImmutableList.copyOf(archives) : ImmutableList.of();
      this.configs = configs != null ? ImmutableMap.copyOf(configs) : ImmutableMap.of();
    }
  }
}

File: clients/client-python/gravitino/api/job/spark_job_template.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements. See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership. The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License. You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied. See the License for the
# specific language governing permissions and limitations
# under the License.

from copy import deepcopy
from typing import List, Dict, Optional, Any
from .job_template import JobTemplate, JobType


class SparkJobTemplate(JobTemplate):
    """
    Represents a job template for executing Spark jobs.
    """

    def __init__(self, builder: "SparkJobTemplate.Builder"):
        """
        Initializes a SparkJobTemplate instance.
        """
        super().__init__(builder)
        self._class_name = builder.class_name
        self._jars = builder.jars
        self._files = builder.files
        self._archives = builder.archives
        self._configs = builder.configs

    @property
    def class_name(self) -> str:
        """
        Returns the class name of the Spark job.

        Returns:
            the class name as a string.
        """
        return self._class_name

    @property
    def jars(self) -> List[str]:
        """
        Returns the list of JAR files associated with this Spark job template.

        Returns:
            List of JAR file paths.
        """
        return self._jars

    @property
    def files(self) -> List[str]:
        """
        Returns the list of files associated with this Spark job template.

        Returns:
            List of file paths.
        """
        return self._files

    @property
    def archives(self) -> List[str]:
        """
        Returns the list of archives associated with this Spark job template.

        Returns:
            List of archive file paths.
        """
        return self._archives

    @property
    def configs(self) -> Dict[str, str]:
        """
        Returns the configuration settings for this Spark job template.

        Returns:
            A dictionary of configuration key-value pairs.
        """
        return self._configs

    def job_type(self) -> JobType:
        """
        Returns the type of the job, which is SPARK for this template.

        Returns:
            JobType.SPARK: Indicates that this is a Spark job template.
        """
        return JobType.SPARK

    @staticmethod
    def builder() -> "SparkJobTemplate.Builder":
        """
        Creates a new builder instance for constructing a SparkJobTemplate.

        Returns:
            SparkJobTemplate.Builder: A new builder instance.
        """
        return SparkJobTemplate.Builder()

    def __eq__(self, other: Any) -> bool:
        if not super().__eq__(other):
            return False
        return (
            self._class_name == other.class_name
            and self._jars == other.jars
            and self._files == other.files
            and self._archives == other.archives
            and self._configs == other.configs
        )

    def __hash__(self) -> int:
        return hash(
            (
                super().__hash__(),
                self._class_name,
                tuple(self._jars),
                tuple(self._files),
                tuple(self._archives),
                frozenset(self._configs.items()),
            )
        )

    def __str__(self) -> str:
        result = "\nSparkJobTemplate{\n"
        result += f"  className='{self.class_name}',\n"

        if self._jars:
            result += (
                "  jars=[\n" + "".join(f"    {j},\n" for j in self._jars) + "  ],\n"
            )
        else:
            result += "  jars=[],\n"

        if self._files:
            result += (
                "  files=[\n" + "".join(f"    {f},\n" for f in self._files) + "  ],\n"
            )
        else:
            result += "  files=[],\n"

        if self._archives:
            result += (
                "  archives=[\n"
                + "".join(f"    {a},\n" for a in self._archives)
                + "  ],\n"
            )
        else:
            result += "  archives=[],\n"

        if self._configs:
            result += (
                "  configs={\n"
                + "".join(f"    {k}: {v},\n" for k, v in self.configs.items())
                + "  },\n"
            )
        else:
            result += "  configs={},\n"

        return result + super().__str__() + "}\n"

    class Builder(
        JobTemplate.BaseBuilder["SparkJobTemplate.Builder", "SparkJobTemplate"]
    ):
        """
        Builder class for creating SparkJobTemplate instances.
        """

        def __init__(self):
            """
            Initializes the builder with default values.
            """
            super().__init__()
            self.class_name: Optional[str] = None
            self.jars: List[str] = []
            self.files: List[str] = []
            self.archives: List[str] = []
            self.configs: Dict[str, str] = {}

        def with_class_name(self, class_name: str) -> "SparkJobTemplate.Builder":
            """
            Sets the class name for this Spark job template.

            Args:
                class_name: The fully qualified name of the Spark job class.

            Returns:
                The builder instance for method chaining.
            """
            self.class_name = class_name
            return self

        def with_jars(self, jars: List[str]) -> "SparkJobTemplate.Builder":
            """
            Sets the JAR files for this Spark job template.

            Args:
                jars: A list of JAR file paths to be included in the Spark job.

            Returns:
                The builder instance for method chaining.
            """
            self.jars = deepcopy(jars)
            return self

        def with_files(self, files: List[str]) -> "SparkJobTemplate.Builder":
            """
            Sets the files for this Spark job template.

            Args:
                files: A list of file paths to be included in the Spark job.

            Returns:
                The builder instance for method chaining.
            """
            self.files = deepcopy(files)
            return self

        def with_archives(self, archives: List[str]) -> "SparkJobTemplate.Builder":
            """
            Sets the archives for this Spark job template.

            Args:
                archives: A list of archive file paths to be included in the Spark job.

            Returns:
                The builder instance for method chaining.
            """
            self.archives = deepcopy(archives)
            return self

        def with_configs(self, configs: Dict[str, str]) -> "SparkJobTemplate.Builder":
            """
            Sets the configuration settings for this Spark job template.

            Args:
                configs: A dictionary of configuration key-value pairs to be applied to the Spark job.

            Returns:
                The builder instance for method chaining.
            """
            self.configs = deepcopy(configs)
            return self

        def validate(self):
            """
            Validates the SparkJobTemplate properties.
            """
            super().validate()
            if not self.class_name or not self.class_name.strip():
                raise ValueError("Class name must not be null or empty")
            self.jars = self.jars or []
            self.files = self.files or []
            self.archives = self.archives or []
            self.configs = self.configs or {}

        def build(self) -> "SparkJobTemplate":
            """
            Builds and returns a SparkJobTemplate instance.
            """
            self.validate()
            return SparkJobTemplate(self)
